hogben/tests/test_fisher.py

```diff
@@ -0,0 +1,337 @@
+    np.testing.assert_allclose(g_reference, g_correct, rtol=1e-08)
+
+
+@patch('hogben.utils.reflectivity')
+@pytest.mark.parametrize('model_class', ("mock_refl1d_model",
+                                         "mock_refnx_model"))
+def test_fisher_importance_correct_values(mock_reflectivity, model_class,
+                                          request):
+    """
+    Tests that the values of the calculated Fisher information matrix
+    are calculated correctly when an importance scaling is applied.
+    """
+    model = request.getfixturevalue(model_class)
+    xi = model.xi[:3]
+    for index, param in enumerate(xi):
+        param.importance = index + 1
+    mock_reflectivity.side_effect = get_mock_reflectivity()
+    g_correct = [
+        [1.28125, 1.025, 76.875],
+        [0.5125, 0.41, 30.75],
+        [25.625, 20.5, 1537.5]
+    ]
+    g_reference = get_fisher_information([model], xi=xi)
+    np.testing.assert_allclose(g_reference, g_correct, rtol=1e-08)
+
+
+@pytest.mark.parametrize('model_class', ("refl1d_model",
+                                         "refnx_model"))
+@pytest.mark.parametrize('step', (0.01, 0.0075, 0.0025, 0.001, 0.0001))
+def test_fisher_consistent_steps(step, model_class, request):
+    """
+    Tests whether the Fisher information remains mostly consistent when
+    changing step size using the refnx model
+    """
+    model = request.getfixturevalue(model_class)
+    g_reference = get_fisher_information([model], step=0.005)
+    g_compare = get_fisher_information([model], step=step)
+    np.testing.assert_allclose(g_reference, g_compare, rtol=1e-02, atol=0)
+
+
+@patch('hogben.utils.reflectivity')
+@pytest.mark.parametrize('model_class', ("mock_refl1d_model",
+                                         "mock_refnx_model"))
+@pytest.mark.parametrize('model_params', (1, 2, 3, 4))
+def test_fisher_shape(mock_reflectivity, model_params, model_class, request):
+    """
+    Tests whether the shape of the Fisher information matrix remains
+     correct when changing the amount of parameters
+    """
+    model = request.getfixturevalue(model_class)
+    xi = model.xi[:model_params]
+
+    mock_reflectivity.side_effect = get_mock_reflectivity()
+
+    expected_shape = (model_params, model_params)
+    g = get_fisher_information([model], xi=xi)
+    np.testing.assert_array_equal(g.shape, expected_shape)
+
+
+@patch('hogben.utils.reflectivity')
+@pytest.mark.parametrize('model_class', ("mock_refl1d_model",
+                                         "mock_refnx_model"))
+@pytest.mark.parametrize('qs',
+                         (np.arange(0.001, 1.0, 0.25),
+                          np.arange(0.001, 1.0, 0.10),
+                          np.arange(0.001, 1.0, 0.05),
+                          np.arange(0.001, 1.0, 0.01)))
+def test_fisher_diagonal_positive(mock_reflectivity, qs, model_class, request):
+    """Tests whether the diagonal values in the Fisher information matrix
```

Comment:
My thinking was that the limit goes to zero, but never approaches it. A zero value of the FIM would basically resemble infinite certainty about a parameter, right? Which would be nonphysical, it would require an infinite shift in reflectivity for the step in the gradient matrix.  Which is why I thought I'd test for positives, instead of non-negatives.

Reply:
If your parameterisation of the problem has been done in such a way that none of your measured points depend on the parameters you have chosen then it is completely fine to have zero information about those parameters. think about points taken below the critical edge: they have no information about the SiOx layer thickness...